core/tests/scheduler_cost_adjustment.rs

```diff
@@ -0,0 +1,559 @@
+            TestResult {
+                cost_adjustment: cu_limit as i64 - tx_execution_cost as i64,
+                execution_status: Ok(()),
+            },
+        ),
+        // pre #3799:
+        // Cost model: ignores requested CU limit because no bpf instruction, instead
+        //   reserve CUs for 1 ATL instruction and 1 Compute Budget instruction,
+        // Compute budget: allocate reuested CUs,
+        // VM Execution: consume CUs for 1 ATL and 3 System (CPI-ed 3 times) and 1 Compute Budget, then succeed
+        // Result: adjustment = (ATL + CB) - (ATL + 3 * System + CB) = -3 * 150 = -450
+        //   Note negative adjustment means adjust-up
+        (
+            false,
+            TestResult {
+                cost_adjustment: -3
+                    * solana_system_program::system_processor::DEFAULT_COMPUTE_UNITS as i64,
+                execution_status: Ok(()),
+            },
+        ),
+    ] {
+        assert_eq!(
+            expected,
+            test_setup.execute_test_transaction(
+                &[
+                    test_setup.create_lookup_table_ix(),
+                    test_setup.set_cu_limit_ix(cu_limit),
+                ],
+                is_simd_170_enabled
+            )
+        );
+    }
+}
+
+#[test]
+fn test_builtin_ix_set_cu_price_only() {
+    let mut test_setup = TestSetup::new();
+    let mut cu_price = 1;
+
+    // single Compute Budget instruction to set CU Price
+    for (is_simd_170_enabled, expected) in [
+        // post #3799:
+        // Cost model & Compute budget: reserve/allocate default CU for one builtin ix
+        // VM Execution: consume CUs for 1 Compute Budget, then succeed
+        // Result: adjustment = 3_000 - 150 = 2_850
+        (
+            true,
+            TestResult {
+                cost_adjustment: MAX_BUILTIN_ALLOCATION_COMPUTE_UNIT_LIMIT as i64
+                    - solana_compute_budget_program::DEFAULT_COMPUTE_UNITS as i64,
+                execution_status: Ok(()),
+            },
+        ),
+        // pre #3799:
+        // Cost model: reserved CU for 1 compute budget instruction,
+        // Compute budget: allocate zero CU because there is zero non-compute-budget instructions,
+        // VM Execution: no budget to execute, fail.
+        // Result: adjustment = 150 - 0 = 150
+        (
+            false,
+            TestResult {
+                cost_adjustment: solana_compute_budget_program::DEFAULT_COMPUTE_UNITS as i64,
+                execution_status: Err(TransactionError::InstructionError(
+                    0,
+                    InstructionError::ComputationalBudgetExceeded,
+                )),
+            },
+        ),
+    ] {
+        assert_eq!(
+            expected,
+            test_setup.execute_test_transaction(
+                &[test_setup.set_cu_price_ix(cu_price)],
+                is_simd_170_enabled
+            )
+        );
+        cu_price += 1;
+    }
+}
+
+#[test]
+fn test_builtin_ix_precompiled() {
```

Comment:
precompiles are accidentally handled more correctly in Before-3799 version. Document it for now, address it later